text/0000-cargo-rustup-discovery.md

```diff
@@ -0,0 +1,482 @@
+[Administrators Group]: https://docs.microsoft.com/en-us/windows/security/identity-protection/access-control/local-accounts#administrator-account
+
+## Implementation details
+
+The Cargo and Rustup teams are working together to implement this change.
+They will endeavor to share the validation logic as much as possible.
+
+
+# Drawbacks, alternatives, and other considerations
+
+The following sections discuss alternatives considered, drawbacks to this proposal,
+and other considerations for why this approach is being considered and possible future changes.
+
+## Comparison to git
+
+This design is largely based on git's behavior.
+However, there are several differences:
+
+* Git restricts the safe directory comparison to only the given directory,
+  and none of its subdirectories.
+  This proposal distinctly deviates from that by allowing subdirectories.
+  It is not clear why git did not implement that behavior.
+  This was chosen because:
+
+    * Cargo projects and configs are nested, and Cargo needs to be able to read all of them.
+      Specifying individual directories could be cumbersome,
+      or more complex logic like "honor a match in the root of a workspace to apply to all members" could be more difficult to implement.
+    * Users may choose to trust their entire home directory (or some subset of it).
+      This can be helpful if the user needs to regularly run tools as a different user across different projects.
+
+  Unfortunately this does have a greater security exposure,
+  but I think the convenience is on balance with the potential risk.
+
+  Alternatively, gitignore-style patterns could be used (like `/home/eric/**`),
+  however we would prefer to avoid the additional complexity.
+
+* Git has the behavior that an empty entry in the safe directories list will "clear" it and ignore all previous entries.
+  This is one way to override the system config.
+
+  Cargo doesn't have system configs, so it may not be as useful.
+
+  One possible use case is if you have `*` in your cargo home config,
+  and you want to set an environment variable to override it.
+
+* Git supports path interpolation in its paths, such as `~` expanding to the user's home directory.
+  At this time, Cargo does not do interpolation for paths,
+  but something like this could be considered.
+  There is a hazard not supporting it in the outset that it could be technically breaking to add it in the future, but I consider that risk extremely low.
+
+* Cargo uses `safe.directories` with a plural unlike git which uses `safe.directory`.
+  Cargo typically uses plurals for arrays (such as "rustflags"),
+  and I think it better conveys that it takes multiple values
+  (particularly in the environment variable `CARGO_SAFE_DIRECTORIES`).
+  However, this difference may cause confusion for users.
+
+## Usability hazards
+
+There are quite a few scenarios where this will cause a false-positive, likely causing frustration for users.
+Some examples:
```

Comment:
Another one is creating throwaway projects in /tmp, right? I often do this myself, but there are also many tools that do so internally, like evcxr and cargo-outdated.

Reply:
I'm not sure I follow.  If you, logged in as yourself, go to `/tmp` and do `cargo new foo` and then `cd foo; cargo run`, that should be fine. That will be owned by you.  Can you explain more?

Reply:
I don't remember what I was thinking about when I commented.